# Chapter 4 - Exercise 1: Thực hiện những yêu cầu liên quan đến series, và đối chiếu với kết quả Output

## Part 1: Thực hiện các phép toán trên series

In [1]:
import numpy as np
import pandas as pd

In [20]:
# Câu 1a: Tạo arr_1 có 5 phần tử số nguyên chẵn 2-4-6-8-10 , arr_2 có 5 phần tử số nguyên lẻ 1-3-5-7-11
arr_1 = np.array([2, 4, 6, 8, 10])
arr_2 = np.array([1, 3, 5, 7, 11])

# In danh sách các phần tử của arr_1 và arr_2
ser1 = pd.Series(arr_1)
ser2 = pd.Series(arr_2)
print (ser1)
print (ser2)

0     2
1     4
2     6
3     8
4    10
dtype: int32
0     1
1     3
2     5
3     7
4    11
dtype: int32


<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  
  <pre>0     2
1     4
2     6
3     8
4    10
dtype: int32

0     1
1     3
2     5
3     7
4    11
dtype: int32
</pre>
</details>

In [3]:
# Câu 1b: Thực hiện phép toán và thể hiện kết quả của:  ser1 + ser2 

print(ser1 + ser2)

0     3
1     7
2    11
3    15
4    21
dtype: int32


<details>
  <summary>Nhấn vào đây để xem kết quả !</summary>
<pre>0     3
1     7
2    11
3    15
4    21
dtype: int32</pre>

</details>

In [4]:
# Câu 1c: Thực hiện phép toán và thể hiện kết quả của:  ser1 - ser2
print (ser1 - ser2)

0    1
1    1
2    1
3    1
4   -1
dtype: int32


In [5]:
# Câu 1d: Thực hiện phép toán và thể hiện kết quả của:  ser1 * ser2
print (ser1 * ser2)

0      2
1     12
2     30
3     56
4    110
dtype: int32


In [6]:
# Câu 1e: Thực hiện phép toán và thể hiện kết quả của:  ser1 / ser2
print (ser1 / ser2)

0    2.000000
1    1.333333
2    1.200000
3    1.142857
4    0.909091
dtype: float64


<details>
  <summary>Nhấn vào đây để xem kết quả !</summary>
<pre>0    2.000000
1    1.333333
2    1.200000
3    1.142857
4    0.909091
dtype: float64</pre>

</details>

In [7]:
# Câu 2a: Kiểm tra xem các phần tử của ser1 có > các phần tử của ser2 không?
print (ser1.gt(ser2))

0     True
1     True
2     True
3     True
4    False
dtype: bool


<details>
  <summary>Nhấn vào đây để xem kết quả !</summary>
<pre>0     True
1     True
2     True
3     True
4    False
dtype: bool</pre>

</details>

In [8]:
# Câu 2b: Kiểm tra xem các phần tử của ser1 có < các phần tử của ser2 không?
print (ser1.lt(ser2))

0    False
1    False
2    False
3    False
4     True
dtype: bool


In [9]:
# Câu 2c: Kiểm tra xem các phần tử của ser1 có = các phần tử của ser2 không?
print (ser1.eq(ser2))

0    False
1    False
2    False
3    False
4    False
dtype: bool


In [21]:
# Câu 3a: Thêm 2 phần tử [6, 12] vào ser2
ser2 = ser2.append(pd.Series([6, 12]))
print(ser2)

0     1
1     3
2     5
3     7
4    11
0     6
1    12
dtype: int64


<details>
  <summary>Nhấn vào đây để xem kết quả !</summary>
<pre>0     1
1     3
2     5
3     7
4    11
0     6
1     12
dtype: int64
</pre>
</details>

In [28]:
# Câu 3b: Tạo series ser3 chỉ chứa các phần tử có trong ser1 mà không có trong ser2. 
#         In danh sách các phần tử của 
ser2.reset_index(inplace=True, drop=True)
ser3 = ser1[~ser1.isin(ser2)]
print(ser3)

0     2
1     4
3     8
4    10
dtype: int32


<details>
  <summary>Nhấn vào đây để xem kết quả !</summary>
<pre>0     2
1     4
3     8
4    10
dtype: int32
</pre>

</details>

In [ ]:
# Câu 3c: Tạo series ser4 chỉ chứa các phần tử có trong ser2 mà không có trong ser1. 
#         In danh sách các phần tử của ser4

<details>
  <summary>Nhấn vào đây để xem kết quả !</summary>
<pre>0     1
1     3
2     5
3     7
4    11
1    12
dtype: int64
</pre>

</details>

In [ ]:
# Câu 4: Tạo series ser5 chứa các phần tử chỉ có trong ser1 và chỉ có trong ser2
#        In danh sách các phần tử của ser5

<details>
  <summary>Nhấn vào đây để xem kết quả !</summary>
<pre>0     2
1     4
3     8
4    10
0     1
1     3
2     5
3     7
4    11
1    12
dtype: int64
</pre>

</details>

## Part 2: Truy xuất các phần tử, và thống kê thông tin trên series

In [ ]:
np.random.seed(1)
# Câu 1a: Tạo series ser6 có 35 phần tử số nguyên ngẫu nhiên có giá trị trong khoảng từ 1 đến 9. 
ser6 = pd.Series(np.random.randint(1, 9, size=35))
print(ser6)

<details>
  <summary>Nhấn vào đây để xem kết quả !</summary>
<pre>(35,)

0    6
1    9
2    6
3    1
4    1
dtype: int32

30    8
31    4
32    7
33    6
34    2
dtype: int32
</pre>

</details>

In [ ]:
# Câu 1b: In danh sách các phần tử của ser6 theo dạng array
print(ser6.values)

<details>
  <summary>Nhấn vào đây để xem kết quả !</summary>
<pre>[6 9 6 1 1 2 8 7 3 5 6 3 5 3 5 8 8 2 8 1 7 8 7 2 1 2 9 9 4 9 8 4 7 6 2]
</pre>

</details>

In [ ]:
# Câu 1c: Cho biết thông tin thống kê chung (describe()) của ser6
ser6.describe()

<details>
  <summary>Nhấn vào đây để xem kết quả !</summary>
<pre>count    35.000000
mean      5.200000
std       2.763204
min       1.000000
25%       2.500000
50%       6.000000
75%       8.000000
max       9.000000
dtype: float64</pre>

</details>

In [ ]:
# Câu 1d: Cho biết tổng của các phần tử có trong ser6
print(ser6.sum())

<details>
  <summary>Nhấn vào đây để xem kết quả !</summary>
<pre> 182 </pre>
</details>

In [ ]:
# Câu 1e: Cho biết phần tử có tần suất xuất hiện nhiều nhất trong ser6

<details>
  <summary>Nhấn vào đây để xem kết quả !</summary>
<pre>0    8
dtype: int32</pre>

</details>

In [ ]:
# Câu 2:  Liệt kê các dòng trong ser6 mà giá trị chia hết cho 2 và cho 3
print(ser6[(ser6 % 2 == 0) & (ser6 % 3 == 0)])

<details>
  <summary>Nhấn vào đây để xem kết quả !</summary>
<pre>0     6
2     6
10    6
33    6
dtype: int32</pre>

</details>

In [ ]:
# Câu 3: In các phần tử ở vị trí 0, 5, 10, 15 trong ser6
print(ser6[:20:5])
print(ser6[[0,6,10,15]])

<details>
  <summary>Nhấn vào đây để xem kết quả !</summary>
<pre>0     6
5     2
10    6
15    8
dtype: int32</pre>

</details>

In [ ]:
# Câu 4: In ra các giá trị unique (array) trong ser6
ser6.unique()

<details>
  <summary>Nhấn vào đây để xem kết quả !</summary>
<pre>array([6, 9, 1, 2, 8, 7, 3, 5, 4], dtype=int64)</pre>
</details>

In [ ]:
# Câu 5: Tạo series ser7 với mỗi phần tử có giá trị = lập phương của phần tử trong ser6. 
ser7 = ser6.map(lambda x: x^3)
print(ser7)

<details>
  <summary>Nhấn vào đây để xem kết quả !</summary>
<pre>0    216.0
1    729.0
2    216.0
3      1.0
4      1.0
dtype: float64
</pre>

</details>

## Part 3: Tạo series từ list, chuỗi và biểu thức điều kiện 

Hướng dẫn chung:
* Lập biểu thức điều kiện: <span style="color:blue">biến_series.map(lambda)</span>

In [ ]:
# Câu 1: Cho list sau:
lst = ["abc", "defg", "htlmj", "dfg", "ljsac"]

In [ ]:
# Câu 1: Tạo series từ list này và tạo series ser_dodai với mỗi phần tử có giá trị là chiều dài của mỗi phần tử trong ser_chuoi
ser_chuoi = pd.Series(lst)
ser_dodai = pd.Series(ser_chuoi.map(lambda x: len(x)))
print(ser_dodai)

<details>
  <summary>Nhấn vào đây để xem kết quả !</summary>
<pre>0    3
1    4
2    5
3    3
4    5
dtype: int64</pre>

</details>

In [33]:
# Câu 2: Cho ser = pd.Series(np.array([1, 2, 4, 5, 8, 7, 6, 9])). 
# Sử dụng biểu thức điều kiện thích hợp để in ra các dòng trong ser có giá trị là số nguyên tố
def isPrime(num):
    if num > 1:
        for i in range(2, num//2):
            if (num % i) == 0:
                return False
                break
            else:
                return True
    else:
        return False
    
ser = pd.Series(np.array([1, 2, 4, 5, 8, 7, 6, 9]))
# print(ser[isPrime(ser)])

TypeError: 'function' object is not iterable

<details>
  <summary>Nhấn vào đây để xem kết quả !</summary>
<pre>1    2
3    5
5    7
dtype: int32
</pre>

</details>

In [ ]:
# Câu 3: Cho mẫu email như sau:
pattern ='[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\\.[A-Za-z]{2,4}'

# Tạo một series ser_ch, với mỗi phần tử trong ser_ch là một chuỗi
# Gợi ý: 'reading newspaper from tuoitre.vn', 'tubirona@gmail.com', 'nguyen.nn@yahoo.com', 'tran_2014@hotmail.com.vn'

# In ra những dòng trong ser_ch thỏa điều kiện chuỗi là email

<details>
  <summary>Nhấn vào đây để xem kết quả !</summary>
<pre>1          tubirona@gmail.com
2         nguyen.nn@yahoo.com
3    tran_2014@hotmail.com.vn
dtype: object</pre>

</details>

In [ ]:
# Câu 4: Cho series:
ser_names = pd.Series(['Manufacturer', 'Model', 'CarType', 'Min_Price', 'Price', 'Max_Price',
       'MPG_city', 'MPG_highway', 'AirBags', 'DriveTrain', 'Cylinders',
       'EngineSize', 'Horsepower', 'RPM', 'Rev_per_mile', 'Man_trans_avail',
       'Fuel_tank_capacity', 'Passengers', 'Length', 'Wheelbase', 'Width',
       'Turn_circle', 'Rear_seat_room', 'Luggage_room', 'Weight', 'Origin',
       'Make'])
# Sử dụng biểu thức điều kiện thích hợp để in ra các dòng của ser_names thỏa điều kiện trong chuỗi có chữ 'Price'


<details>
  <summary>Nhấn vào đây để xem kết quả !</summary>
<pre>3    Min_Price
4        Price
5    Max_Price
dtype: object
</pre>

</details>